[UIK-2426][bulk-textarea] new bulk textarea component (#1893)
semrush/intergalactic: website/docs/components/bulk-textarea/bulk-textarea.md

---
title: BulkTextarea
tabs: Design('bulk-textarea'), A11y('bulk-textarea-a11y'), API('bulk-textarea-api'), Example('bulk-textarea-code'), Changelog('bulk-textarea-changelog')
---

## Description

**BulkTextarea** is a multiline text field designed for entering lists of comma-separated values.

You can enter a list manually using **Enter** or a comma, or paste a comma-separated list. You can also define your own rules for splitting values into rows.

This component also provides the possibility to validate every row and the whole input.

## Component composition

> ![](static/bulktextarea-composition.png)

Reviewer: shifted a bit
<img width="396" alt="image" src="https://github.com/user-attachments/assets/da519bae-f4d7-4e6e-8b8f-057036a751dd" />

Component consists of the following:

- `BulkTextarea.InputField`
- `BulkTextarea.Counter`
- `BulkTextarea.InputField.Row`
- `BulkTextarea.ErrorItem`
- `BulkTextarea.ErrorsNavigation`
- `BulkTextarea.ClearAllButton`

## Appearance

### Sizes

Component has two sizes:

Table: BulkTextarea sizes

| Size (font size) | Appearance example     |
| ---------------- | ---------------------- |

> | M (`--fs-200`)   | ![](static/bulk-m.png) |

Reviewer: some thoughts about alts for illustrations
we'll have to return to his page again in the task about alt texts, so why not do it right from the start
I don't like the idea of useless alts ("example of textarea of size L"), and empty alt isn't good either in this case, because it creates an empty cell for SR
I tried restructuring this table into a bullet list, or merging columns, but it doesn't look good

so I have another idea
what if we do something like this:
<img width="431" alt="image" src="https://github.com/user-attachments/assets/f7fcf471-95ac-440f-92c9-3c3da479e3a1" />
this way we also fix another problem — different sizes scaled unconsistently on different illustrations, on one illustration all sizes will look correctly relative to each other

| L (`--fs-300`)   | ![](static/bulk-l.png) |

> ### Rows

Reviewer: I think it should be called "lines" in docs and in component/prop names, left a message in chat to discuss

Each row is numbered. Rows that fail validation are marked with a `Warning` icon on the right. Rows that exceed the limit are highlighted.

![](static/bulktextarea-rows.png)

### Width and height

Avoid making the textarea smaller than 160–200px in width and 3–4 rows in height. Working with large amounts of data in smaller sizes can be challenging, especially when it serves as a primary input in a form.

## Counter

BulkTextarea includes a counter displaying the number of rows entered. It's always positioned next to the text label.

Table: BulkTextarea with counter

| Size (font size) | Input with label       |
| ---------------- | ---------------------- |
| M (`--fs-200`)   | ![](static/bulk-m.png) |
| L (`--fs-300`)   | ![](static/bulk-l.png) |

### How counter works

1. The counter shows the value limit and updates once user enters at least one character on the line.
2. The counter updates if the user deletes a character and a line.
3. If the user presses **Enter** or types a comma on an empty line, they stay on the same line. This helps avoid adding unnecessary blank lines.
4. For now duplicate lines are counted.

### Values near and above limit

> When the number of values reaches the limit, the counter changes its theme to `warning`.

Reviewer: (orrr maybe not, I'm not 100% sure, for me "values" is at least clearer than "rows", but I like "lines" most of course)

![](static/values-limit.png)

> Rows with values that exceed the limit are highlighted. Counter changes its theme to `danger`.

Reviewer: ```suggestion
Lines that exceed the limit are highlighted. Counter changes its theme to `danger`.
```
values as well

![](static/values-off-limit.png)

## Controls

In the component, the controls have a default placement, but if you need to position them elsewhere, you can expand the component and arrange them differently. [Refer to the example](/components/bulk-textarea/bulk-textarea-code#expanded-bulktextarea).

> ### Clear field button

Reviewer: ```suggestion
### Clear all button
```
?

As soon as at least one character is entered in the field, a button for clearing the field appears next to it. By default, the button is positioned to the bottom right of the field.

![](static/bulktextarea-clear-button.png)

### Navigation buttons

Error navigation shows up after all the values or a single value (depending on the validation method) are validated. The number of errors is shown next to the buttons.

![](static/bulktextarea-error-navigation.png)

> When a button is clicked, the cursor moves to the first or last invalid row, depending on the button. The entire row's value is selected for quick deletion. The text next to the navigation buttons updates to _"n out of n"_ when the cursor is moved using the buttons, keyboard, or mouse.

Reviewer: ```suggestion
When a button is pressed, the cursor moves to the first or last invalid row, depending on the button. The entire row's value is selected for quick deletion. The text next to the navigation buttons updates to _"n out of n"_ when the cursor is moved using the buttons, keyboard, or mouse.
```
a bit more inclusive, I think
